docs/type_inference.md

```diff
@@ -0,0 +1,60 @@
+
+
+## (P2) Make the most of ODS
+
+ODS files (like [StablehloOps.td](https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/StablehloOps.td)) define ops with traits and types for each operands/attributes/results and will do the verifications. Thus NO verification code needed in the verifiers or shape functions for the properties which are already guaranteed by the ODS.  Remove the verification code if duplicated with ODS, as they will never be triggered.
+
+Do we need adding tests for the constraints from the ODS? Please see “P5 Establish testing guidelines” below.
+
+
+## (P3) Merge verifiers into shape functions
+
+Both 
+- **verifiers**: implemented by `Op::verify()`, and 
+- **shape functions**: implemented by `InferTypeOpInterface` like `Op::inferReturnTypes()` or `Op::inferReturnTypeComponents` 
+
+may have verification code to check operands/attributes/results. An ideal split would be that: let the verifiers check the operands/attributes, then let shape functions to calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
+
+1. Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
+2. Maintenance burden: as verifications of an op are contained in two different method. 
+
+Thus, in most cases the verifiers will be removed and all the verification logic will be merged into the shape functions (example: [PR#135](https://github.com/openxla/stablehlo/pull/135)). The only exceptions will be:
+
+1. Op that cannot have a shape function: like `stablehlo.reshape`, whose result type cannot be inferred.
+2. (should be rare) Op that have tons of logic in their verifier but very little logic in their shape function.
+
+
+## (P4) Align error messages by following standard templates. 
+
+Different error messages for similar types of error have inconsistent user experience and more maintenance efforts. We propose using utils to wrap each type of common errors and apply gradually to the existing code.
```

Comment:
Given that we don't have templates yet, let's drop this section.

Reply:
Done.